modules/gapi/src/backends/oak/goakbackend.cpp

```diff
@@ -0,0 +1,616 @@
+                     std::vector<cv::GArg>& oak_wrapped_args);
+    void outArg(const RcDesc &rc, std::vector<ExtractTypeHelper::OutputPtr>& oak_outs);
+
+    const ade::Graph& m_g;
+    GModel::ConstGraph m_gm;
+    cv::GCompileArgs m_args;
+
+    std::unordered_map<ade::NodeHandle,
+                       std::shared_ptr<dai::Node>,
+                       ade::HandleHasher<ade::Node>> m_oak_nodes;
+    std::unordered_set<ade::NodeHandle, ade::HandleHasher<ade::Node>> m_processed_nodes;
+    // Will be reworked later when XLinkIn will be introduced as input
+    std::shared_ptr<dai::node::ColorCamera> m_camera_input;
+    std::tuple<int, int> m_camera_wh;
+
+    std::unordered_map<ade::NodeHandle,
+                       std::vector<ExtractTypeHelper::InputPtr>,
+                       ade::HandleHasher<ade::Node>> m_oak_node_inputs;
+    std::unordered_map<ade::NodeHandle,
+                       std::vector<ExtractTypeHelper::OutputPtr>,
+                       ade::HandleHasher<ade::Node>> m_oak_node_outputs;
+
+    // Backend outputs
+    std::vector<std::shared_ptr<dai::node::XLinkOut>> m_xlink_outputs;
+    std::vector<std::shared_ptr<dai::DataOutputQueue>> m_out_queues;
+    std::vector<std::string> m_out_queue_names;
+
+    // Backend inputs
+    std::vector<std::pair<std::string, dai::Buffer>> m_in_queues;
+
+    // Note: dai::Pipeline should be the only one for the whole pipeline,
+    // so there is no way to insert any non-OAK node in graph between other OAK nodes.
+    // The only heterogeneous case possible is if we insert other backends after or before
+    // OAK island.
+    std::unique_ptr<dai::Device> m_device;
+    std::unique_ptr<dai::Pipeline> m_pipeline;
+
+    std::vector<cv::GArg> m_oak_wrapped_args;
+
+public:
+    GOAKExecutable(const ade::Graph& g,
+                   const cv::GCompileArgs& args,
+                   const std::vector<ade::NodeHandle>& nodes,
+                   const std::vector<cv::gimpl::Data>& ins_data,
+                   const std::vector<cv::gimpl::Data>& outs_data);
+    ~GOAKExecutable() = default;
+
+    // FIXME: could it reshape?
+    virtual bool canReshape() const override { return false; }
+    virtual void reshape(ade::Graph&, const GCompileArgs&) override {
+        util::throw_error
+            (std::logic_error
+             ("GOAKExecutable::reshape() is not supported"));
+    }
+
+    virtual void handleNewStream() override;
+    virtual void handleStopStream() override;
+};
+
+class GOAKContext {
+public:
+    // FIXME: make private?
+    using Input = GOAKExecutable::ExtractTypeHelper::Input;
+    using Output = GOAKExecutable::ExtractTypeHelper::Output;
+    using InputPtr = GOAKExecutable::ExtractTypeHelper::Input*;
+    using OutputPtr = GOAKExecutable::ExtractTypeHelper::Output*;
+
+    GOAKContext(const std::unique_ptr<dai::Pipeline>& pipeline,
+                const std::tuple<int, int>& camera_size,
```

Comment:
why not `cv::Size`? I'd stick to OpenCV types as much as possible here.

Reply:
Done